### PR TITLE
[KeyValueSnapshotReader Phase I] Part 7: ReadSnapshotValidator

### PR DESCRIPTION
**Pull request by jeremyk-91**
## General
**Before this PR**: The validation of read snapshots, in particular @LucasIME 's work on methods of dealing with partial as opposed to guaranteed complete reads, happens largely in `SnapshotTransaction`, which already does many other things.

**After this PR**:

==COMMIT_MSG==
Logic for handling the validation of read snapshots is moved to another class, `ReadSnapshotValidator`.
==COMMIT_MSG==

**Priority**: High P2; no specific priority but the workstream in general is high priority.

**Concerns / possible downsides (what feedback would you like?)**: 
- There is some tricky boolean logic, so the main risk on this one is accidentally getting something wrong. I added tests, of course.
- I don't really like exposing `boolean doesTableRequirePreCommitValidation(TableReference tableRef, boolean allPossibleCellsReadAndPresent)`, though the main alternative (something like `void completeValidationForTables(Set<TableReference> tables, boolean hasPossiblyUnvalidatedReads`) doesn't exactly seem like something a `ReadSnapshotValidator` should do especially given the specific circumstances this is used by (read-write SnapshotTransaction that only did reads).

**Is documentation needed?**: No

## Compatibility
**Does this PR create any API breaks (e.g. at the Java or HTTP layers) - if so, do we have compatibility?**: No

**Does this PR change the persisted format of any data - if so, do we have forward and backward compatibility?**: No

**The code in this PR may be part of a blue-green deploy. Can upgrades from previous versions safely coexist? (Consider restarts of blue or green nodes.)**: Yes

**Does this PR rely on statements being true about other products at a deployment - if so, do we have correct product dependencies on these products (or other ways of verifying that these statements are true)?**: I don't think so

**Does this PR need a schema migration?** No

## Testing and Correctness
**What, if any, assumptions are made about the current state of the world? If they change over time, how will we find out?**: Nothing in particular

**What was existing testing like? What have you done to improve it?**: Added tests for RSV. General behaviour should still be validated by the general SnapshotTransaction tests.

**If this PR contains complex concurrent or asynchronous code, is it correct? The onus is on the PR writer to demonstrate this.**: N/A

**If this PR involves acquiring locks or other shared resources, how do we ensure that these are always released?**: N/A

## Execution
**How would I tell this PR works in production? (Metrics, logs, etc.)**: No changes

**Has the safety of all log arguments been decided correctly?**: N/A

**Will this change significantly affect our spending on metrics or logs?**: No

**How would I tell that this PR does not work in production? (monitors, etc.)**: Failures in behaviour

**If this PR does not work as expected, how do I fix that state? Would rollback be straightforward?**: Rollback

**If the above plan is more complex than “recall and rollback”, please tag the support PoC here (if it is the end of the week, tag both the current and next PoC)**:

## Scale
**Would this PR be expected to pose a risk at scale? Think of the shopping product at our largest stack.**: No

**Would this PR be expected to perform a large number of database calls, and/or expensive database calls (e.g., row range scans, concurrent CAS)?**: No

**Would this PR ever, with time and scale, become the wrong thing to do - and if so, how would we know that we need to do something differently?**: Not that I'm aware of

## Development Process
**Where should we start reviewing?**: DefaultReadSnapshotValidatorTests

**If this PR is in excess of 500 lines excluding versions lock-files, why does it not make sense to split it?**:

**Please tag any other people who should be aware of this PR**:
@jeremyk-91
@sverma30
@raiju

### CODE REVIEW

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/precommit/ReadSnapshotValidator.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+
+public interface ReadSnapshotValidator {
+    ValidationState throwIfPreCommitRequirementsNotMetOnRead(
+            TableReference tableRef, long timestamp, boolean allPossibleCellsReadAndPresent);
+
+    boolean doesTableRequirePreCommitValidation(TableReference tableRef, boolean allPossibleCellsReadAndPresent);
+
+    enum ValidationState {
```
Comment (LucasIME) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/precommit/ReadSnapshotValidator.java:44]: Probably worth adding some docs explaining what the validation means in this case?

Comment (jeremyk-91) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/precommit/ReadSnapshotValidator.java:44]: yep! will do

**Comment by jeremyk-91** [comment]
I need to do checkstyle stuff

**Comment by jeremyk-91** [comment]
oops, something weird happened here. I'll fix this and update here when it's done

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/precommit/ReadSnapshotValidator.java (hunk 1 above)
Comment (jeremyk-91) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/precommit/ReadSnapshotValidator.java:44]: (added to the class and methods, though happy to add more context here if we think it's useful!)

**Review by jeremyk-91 — COMMENTED** [review]
I resolved the merge conflicts/weirdness so this should now be ready for review.

**Review by LucasIME — APPROVED** [review]

**Pull request merged**